crates/core/src/node/inner/blockchain.rs

```diff
@@ -0,0 +1,369 @@
+
+    pub async fn get_block_by_id(
+        &self,
+        block_id: api::BlockId,
+    ) -> Option<api::Block<api::TransactionVariant>> {
+        self.inspect_block_by_id(block_id, |block| block.clone())
+            .await
+    }
+
+    pub async fn get_block_hash_by_number(&self, number: L2BlockNumber) -> Option<H256> {
+        self.inspect_block_by_number(number, |block| block.hash)
+            .await
+    }
+
+    pub async fn get_block_hash_by_id(&self, block_id: api::BlockId) -> Option<H256> {
+        self.inspect_block_by_id(block_id, |block| block.hash).await
+    }
+
+    pub async fn get_block_number_by_hash(&self, hash: &H256) -> Option<L2BlockNumber> {
+        self.inspect_block_by_hash(hash, |block| L2BlockNumber(block.number.as_u32()))
+            .await
+    }
+
+    pub async fn get_block_number_by_id(&self, block_id: api::BlockId) -> Option<L2BlockNumber> {
+        self.inspect_block_by_id(block_id, |block| L2BlockNumber(block.number.as_u32()))
+            .await
+    }
+
+    pub async fn inspect_block_by_hash<T>(
+        &self,
+        hash: &H256,
+        f: impl FnOnce(&api::Block<api::TransactionVariant>) -> T,
+    ) -> Option<T> {
+        Some(f(self.inner.read().await.blocks.get(hash)?))
+    }
+
+    pub async fn inspect_block_by_number<T>(
+        &self,
+        number: L2BlockNumber,
+        f: impl FnOnce(&api::Block<api::TransactionVariant>) -> T,
+    ) -> Option<T> {
+        let storage = self.inner.read().await;
+        let hash = storage.get_block_hash_by_number(number)?;
+        Some(f(storage.blocks.get(&hash)?))
+    }
+
+    pub async fn inspect_block_by_id<T>(
+        &self,
+        block_id: api::BlockId,
+        f: impl FnOnce(&api::Block<api::TransactionVariant>) -> T,
+    ) -> Option<T> {
+        let storage = self.inner.read().await;
+        let hash = storage.get_block_hash_by_id(block_id)?;
+        Some(f(storage.blocks.get(&hash)?))
+    }
+
+    pub async fn get_tx_receipt(&self, tx_hash: &H256) -> Option<api::TransactionReceipt> {
+        self.inspect_tx(tx_hash, |tx| tx.receipt.clone()).await
+    }
+
+    pub async fn get_tx_debug_info(
+        &self,
+        tx_hash: &H256,
+        only_top: bool,
+    ) -> Option<api::DebugCall> {
+        self.inspect_tx(tx_hash, |tx| tx.debug_info(only_top)).await
+    }
+
+    pub async fn inspect_tx<T>(
+        &self,
+        tx_hash: &H256,
+        f: impl FnOnce(&TransactionResult) -> T,
+    ) -> Option<T> {
+        Some(f(self.inner.read().await.tx_results.get(tx_hash)?))
+    }
+
+    // TODO: Seems like a strange pattern to allow this
+    pub async fn inspect_all_txs<T>(
+        &self,
+        f: impl FnOnce(&HashMap<H256, TransactionResult>) -> T,
+    ) -> T {
+        f(&self.inner.read().await.tx_results)
+    }
+}
+
+/// A single-instance writer to blockchain state that is only available to [`super::InMemoryNodeInner`].
+pub(super) struct BlockchainWriter {
+    pub(super) inner: Arc<RwLock<Blockchain>>,
+}
```

Comment:
Can't we express `BlockchainReader` and `BlockchainWriter` as two traits? Actually, probably `BlockchainWriter` is not even needed -- you can expose only `Box<dyn BlockchainReader>` outside of the crate, and rework `Blockchain` as `pub(crate) struct Blockchain { inner: Arc<RwLock<Blockchain>> }`.

With that, hopefully, you will be able to expose all the required mutability methods on the structure itself so that you don't have to leak guards in the interface.

Reply:
Done, I had to move some seemingly API-specific logic into `Blockchain` (to avoid exposing generic `inspect_*` methods that are not object safe). But IMHO worked out pretty okay.

I had to wrap `ReadBlockchain` and `ReadTime` into `Arc<dyn _>` instead of `Box<dyn _>` just so they would be cloneable. Do you think it is worth bothering with `DynClone` to avoid this?